doc/ubimport.1
.\" Man page for ubimport
.\" Use the following command to view man page:
.\"
.\"  tbl ubimport.1 | nroff -man | less
.\"
.TH UBIMPORT 1 "January 20, 2007"
.SH NAME
ubimport \- A tool to convert bibles in Unbound Bible format to MQL for importing into Emdros
.SH SYNOPSIS
\fBubimport\fR [ options ]
.br
.SH DESCRIPTION
\fBubimport\fR is a command-line tool to convert bibles in Unbound
Bible format (http://unbound.biola.edu) to MQL, for easy importing
into Emdros.

.SH OPTIONS
\fBubimport\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database name to use at the top of the MQL.  If not set, the
"CREATE DATABASE ... USE DATABASE" incantation won't be issued.
.TP
.BI \-i\ ,\ \-\-input \ filename
read from file filename. This must be a bible in valid Unbound Bible
format.  The default filename is "-", which means "standard input".
.TP
.BI \-o\ ,\ \-\-output \ filename
dump to file filename. The default is "-", which means "standard output".
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)
.SH AUTHORS
Copyright
.Cr
2001-2006 by Ulrik Petersen (ulrikp@users.sourceforge.net).  Note that
this software is distributed under the GNU GPL.  See the sources for
details.


NEWS
This file is for release-notes for Emdros.

- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries
--------------

For this release, we do not provide Linux binaries.  Instead, we have
ensured that it is easy to build .rpms and .debs from the sources.


To build an RPM or .deb, first install the dependencies mentioned in
the text file called "INSTALL" in the root of the sources.


A) To build an RPM:

rpmbuild -ta emdros-3.4.0.tar.gz


B) To build a .deb:

1) cp /path/to/Emdros/sources/emdros-3.4.0.tar.gz .

2) tar xfvz emdros-3.4.0.tar.gz

3) cd emdros-3.4.0

4) dpkg-buildpackage -rfakeroot

This will build the .deb in ../.


Changes in more detail
----------------------

Release 3.4.0 has the following changes over 3.3.0:

- Build changes:

  - Support for compilation on Android and iOS were added.

  - Non-Unicode (ANSI) builds of the programs are no longer available
    on Windows.

  - On Linux / Unix / Mac OS X, wx/libemdrosguiu is now a static
    library.

  - Out-of-the-box support for building on Debian / Ubuntu /
    derivatives was added:

	$ tar xfzv emdros-x.y.z.tar.gz
	$ cd emdros-x.y.z
	$ dpkg-buildpackage -rfakeroot
	$ cd .. # The .deb package is in this directory
	$ sudo dpkg -i emdros_*.deb # This should install it.

  - The .pro build files for Qt have been updated to Qt 4.8.0, which
    entails dropping support for everything below Mac OS X 10.6 on Qt.

  - It is now possible to do Win64 builds on Windows.  See
    the very top of win32\config.mak for how to do it.

  - It is now possible to build Emdros on Windows with MinGW32,
    including support for wxWidgets.

  - Add a test suite which is run with "make test" in the root
    directory of the sources.

  - Much better configure.ac support for wxWidgets 2.8, 2.9, and 3.0,
    especially on Mac OS X.


- Harvest library changes:

  - Bugs were fixed
  - Speeds were increased
  - The template language was enhanced (see the docs)
  - A new Full Text Search engine, FTS2, was added.

- PHP was added as a supported SWIG binding.

- More support for feature-type "SET OF MONADS" was added:

  - Bugs were fixed
  - In topographic queries, it is now possible to say:

    SELECT ALL OBJECTS WHERE 
    [Clause 
        [Phrase
           // Could also be universe instead of substrate.
           // Could also be part_of instead of overlap.
           // part_of(monads, substrate) is still the default.
           overlap(mymonadsetfeature, substrate)
        ]
    ]

- The GET OBJECTS HAVING MONADS IN statement now supports using a
  different monad set than the default "monads" feature:

  GET OBJECTS HAVING MONADS IN
  { 2-40201, 40203-50000 }
  USING MONAD FEATURE described_SOM
  [Describer]
  GO

- Assorted speed increases in the EMdF / MQL libraries.

- The Penn Treebank converter output was improved.

- SQLite 3 was updated to version 3.8.4.3.

- Fixed a few documentation glitches


Enjoy!


Ulrik Sandborg-Petersen


- *** Version 3.3.0 *** 

As usual, binaries are available for Mac OS X, Windows(R), and Fedora
(15).  

The Windows binaries have support for MySQL, SQLite 2, and SQLite 3.
They are built with Visual Studio Express 2008.

The Mac OS X binaries are Universal binaries running on Mac OS X 10.4
(Tiger), 10.5 (Leopard), and 10.6 (Snow Leopard).  They do not have
support for either MySQL or PostgreSQL; Only SQLite 2 and SQLite 3 are
supported in the Mac OS X binaries.  You can compile the sources with
support for MySQL yourself, though, and possibly also PostgreSQL.

The Fedora binaries come with support for PostgreSQL, MySQL, SQLite 2,
and SQLite 3.

Release 3.3.0 has the following changes over 3.2.0:


src/ubimport.cpp
/*
 * ubimport.cpp
 *
 * A front-end to the Unbound Bible import
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 1/7-2009
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */
/************************************************************************
 *
 *   Emdros - the database engine for analyzed or annotated text
 *   Copyright (C) 2001-2009  Ulrik Sandborg-Petersen
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation, license version 2.  
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 *   02111-1307 USA
 *
 *
 *   Special exception
 *   =================
 * 
 *   In addition, as a special exception, Ulrik Petersen, the
 *   copyright holder of Emdros, gives permission to link Emdros, in
 *   whole or in part, with the libraries which are normally
 *   distributed with:
 *   
 *   a) Sun's Java platform,
 *   b) Python, 
 *   c) Jython,
 *   d) Ruby, and/or 
 *   e) Perl 
 *   f) PostgreSQL
 *   g) OpenSSL
 *
 *   (or with modified versions of these), and to distribute linked
 *   combinations including both Emdros, in whole or in part, and one
 *   or more of the libraries normally distributed with (a)-(g) above.
 *
 *   Please note: This gives you special rights concerning the
 *   libraries which normally accompany the above pieces of software.
 *   It gives you no special rights concerning software that you write
 *   yourself.  You must obey the GNU General Public License in all
 *   respects for all of the code used other than the libraries
 *   normally distributed with (a)-(g) above.
 *
 *   If you modify this file, you may extend this exception to your
 *   version of the file, but you are not obligated to do so. If you
 *   do not wish to do so, delete this exception statement from your
 *   version.
 *
 *
 *   Other licensing forms
 *   =====================
 *
 *   If you wish to negotiate commercial licensing, please contact
 *   Ulrik Petersen at ulrikp[at]users.sourceforge.net.
 *
 *   Licensing can also be negotiated if your organization is an
 *   educational, non-profit, charity, missionary or similar
 *   organization.
 *
 *
 *   Website
 *   =======
 *
 *   Emdros has a website here:
 *
 *   http://emdros.org
 *
 *
 *
 **************************************************************************/


#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc

#include <agxml.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <emdros_environment.h>
#include <encryption.h>

#include <unboundbibleimporter.h>

#include <opt.h>


int exec_stream(std::istream *pInStream, std::ostream* pOutStream, const std::string& dbname, eCharsets charset)
{
	UNUSED(charset); // FIXME: Use it, or remove it from parameters list.
	
	int result = 0;

	UnboundBibleImporter *pImporter = new UnboundBibleImporter();
	if (!pImporter->readDocumentFromStream(pInStream)) {
		std::cerr << "FAILURE: Error occurred reading Bible." << std::endl;
		result = 5;
	} else {
		if (dbname != "") {
			(*pOutStream) << "CREATE DATABASE '" << dbname << "' GO\n"
				      << "USE DATABASE '" << dbname << "' GO\n\n";

		}
		pImporter->putSchema(pOutStream);
		pImporter->putCorpusAsMQL(pOutStream);
	}

	delete pImporter;

	return result;
}

bool open_file_for_writing(std::ofstream& fout, const std::string& filename)
{
	bool bResult = false;
	fout.open(filename.c_str());
	if (!fout) {
		std::cerr << "FAILURE: Could not open file " << filename << " for writing." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

bool open_file_for_reading(std::ifstream& fin, const std::string& filename)
{
	bool bResult = false;
	fin.open(filename.c_str());
	if (!fin) {
		std::cerr << "FAILURE: Could not open file " << filename << " for reading." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}


void show_version(std::ostream& ostr)
{
	ostr << "ubimport from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   ubimport [options]" << std::endl;
	ostr << "   ubimport (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   --help               Show this help" << std::endl;
	ostr << "   -V , --version       Show version" << std::endl;
	ostr << "   -d , --dbname db     database name (to use in MQL output)" << std::endl;
	ostr << "   -i , --input  file   Read Bible from filename file (\"-\" means stdin)" << std::endl;
	ostr << "   -o , --output file   Dump database to filename file (\"-\" means stdout)" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   -i -                 (Use stdin)" << std::endl;
	ostr << "   -o -                 (Use stdout)" << std::endl;
	ostr << "   -d \"\"                (Empty means: Don't emit CREATE/USE DATABASE)" << std::endl;
}


int main(int argc, char* argv[])
{
	// Set defaults
	std::string db("");
	std::string output_filename("-");
	std::string input_filename("-");
	std::string dummy_hostname;
	std::string dummy_user;
	eCharsets dummy_charset = kCSUTF8;
	std::string dummy_password;
	eBackendKind dummy_backend_kind;


	addStandardArguments(false // Do not add -e | --encoding
			     );

	addOption("-d", "--dbname", true, 
		  "", // Default value
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("-i", "--input", true, 
		  "-", // Default value
		  "ERROR: -i and --input must have a filename (or \"-\" for stdin)\n"
		  "       as their argument.\n"
		  "       example: -i danish1933.txt\n");

	addOption("-o", "--output", true, 
		  "-", // Default value
		  "ERROR: -o and --output must have a filename (or \"-\" for stdout)\n"
		  "       as their argument.\n"
		  "       example: -o output.mql\n");


	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  dummy_hostname,
					  dummy_user,
					  dummy_password,
					  dummy_backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// --dbname | -d
		getArgumentValue("-d", db);

		// Input
		getArgumentValue("-i", input_filename);

		// Output
		getArgumentValue("-o", output_filename);

		if (surplus_arguments.size() != 0) {
			print_usage(std::cerr);
			std::cerr << "ERROR: Extraneous argument(s) not parsed: '" << joinList("' and '", surplus_arguments) << "'." << std::endl;
			return 1;
		}
	}

	int nResult = 3;
	bool bContinue = true;
	try {
		std::ostream *pOut = 0;
		std::ofstream fout;
		if (output_filename == "-") {
			pOut = &std::cout;
		} else {
			if (!open_file_for_writing(fout, output_filename)) {
				nResult = 4;
				bContinue = false;
			} else {
				pOut = &fout;
			}
		}

		std::istream *pIn = 0;
		std::ifstream fin;
		if (bContinue) {
			if (input_filename == "-") {
				pIn = &std::cin;
			} else {
				if (!open_file_for_reading(fin, input_filename)) {
					nResult = 4;
					bContinue = false;
				} else {
					pIn = &fin;
				}
				
			}
		}

		if (bContinue) {
			nResult = exec_stream(pIn, pOut, db, dummy_charset);
		}

	} catch (EMdFNULLValueException e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	//	delete pEE;
	return nResult;
}


